File: desktop/windows/src/key_spec.cc
#include "key_spec.h"

#include <algorithm>
#include <cctype>

namespace voicestick {
namespace {

std::string Trim(const std::string& s) {
    size_t begin = 0;
    while (begin < s.size() && std::isspace(static_cast<unsigned char>(s[begin]))) ++begin;
    size_t end = s.size();
    while (end > begin && std::isspace(static_cast<unsigned char>(s[end - 1]))) --end;
    return s.substr(begin, end - begin);
}

std::string Lower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(),
                   [](char c) {
                       return static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
                   });
    return s;
}

// 主键名 → VK。未知名返回 nullopt。
std::optional<UINT> MainKeyVkey(const std::string& lower) {
    if (lower.size() == 1) {
        const char ch = static_cast<char>(std::toupper(static_cast<unsigned char>(lower[0])));
        if (ch >= 'A' && ch <= 'Z') return static_cast<UINT>(ch);
        if (ch >= '0' && ch <= '9') return static_cast<UINT>(ch);
        return std::nullopt;
    }
    if (lower == "space") return VK_SPACE;
    if (lower == "enter" || lower == "return") return VK_RETURN;
    if (lower == "esc" || lower == "escape") return VK_ESCAPE;
    if (lower == "tab") return VK_TAB;
    if (lower == "backspace") return VK_BACK;
    if (lower == "delete") return VK_DELETE;
    if (lower == "insert") return VK_INSERT;
    if (lower == "up") return VK_UP;
    if (lower == "down") return VK_DOWN;
    if (lower == "left") return VK_LEFT;
    if (lower == "right") return VK_RIGHT;
    if (lower == "pageup") return VK_PRIOR;
    if (lower == "pagedown") return VK_NEXT;
    if (lower == "home") return VK_HOME;
    if (lower == "end") return VK_END;
    if (lower == "volumeup") return VK_VOLUME_UP;
    if (lower == "volumedown") return VK_VOLUME_DOWN;
    if (lower == "volumemute") return VK_VOLUME_MUTE;
    if (lower[0] == 'f' && lower.size() >= 2) {
        int num = 0;
        for (size_t i = 1; i < lower.size(); ++i) {
            if (lower[i] < '0' || lower[i] > '9') return std::nullopt;
            num = num * 10 + (lower[i] - '0');
            if (num > 24) return std::nullopt;  // 防 int 溢出
        }
        if (num >= 1 && num <= 24) return VK_F1 + (num - 1);
    }
    return std::nullopt;
}

std::string VkeyDisplayName(UINT vk) {
    if (vk >= 'A' && vk <= 'Z') return std::string(1, static_cast<char>(vk));
    if (vk >= '0' && vk <= '9') return std::string(1, static_cast<char>(vk));
    switch (vk) {
        case VK_SPACE: return "Space";
        case VK_RETURN: return "Enter";
        case VK_ESCAPE: return "Esc";
        case VK_TAB: return "Tab";
        case VK_BACK: return "Backspace";
        case VK_DELETE: return "Delete";
        case VK_INSERT: return "Insert";
        case VK_UP: return "Up";
        case VK_DOWN: return "Down";
        case VK_LEFT: return "Left";
        case VK_RIGHT: return "Right";
        case VK_PRIOR: return "PageUp";
        case VK_NEXT: return "PageDown";
        case VK_HOME: return "Home";
        case VK_END: return "End";
        case VK_VOLUME_UP: return "VolumeUp";
        case VK_VOLUME_DOWN: return "VolumeDown";
        case VK_VOLUME_MUTE: return "VolumeMute";
        default: break;
    }
    if (vk >= VK_F1 && vk <= VK_F24) return "F" + std::to_string(vk - VK_F1 + 1);
    return {};
}

} // namespace

std::optional<KeySpec> ParseKeySpec(const std::string& text) {
    KeySpec spec;
    bool ctrl = false, alt = false, shift = false, win = false;
    bool has_main = false;

    size_t pos = 0;
    const std::string input = Trim(text);
    while (true) {
        const size_t plus = input.find('+', pos);
        const std::string part = Lower(Trim(input.substr(
            pos, plus == std::string::npos ? std::string::npos : plus - pos)));
        if (part.empty()) return std::nullopt;

        if (part == "ctrl" || part == "control") {
            if (ctrl) return std::nullopt;
            ctrl = true;
        } else if (part == "alt") {
            if (alt) return std::nullopt;
            alt = true;
        } else if (part == "shift") {
            if (shift) return std::nullopt;
            shift = true;
        } else if (part == "win" || part == "windows" || part == "meta") {
            if (win) return std::nullopt;
            win = true;
        } else {
            if (has_main) return std::nullopt;  // 多个主键
            const auto vk = MainKeyVkey(part);
            if (!vk.has_value()) return std::nullopt;
            spec.vk = *vk;
            has_main = true;
        }
        if (plus == std::string::npos) break;
        pos = plus + 1;
    }
    if (!has_main) return std::nullopt;  // 仅修饰键

    // 修饰键固定 Ctrl/Alt/Shift/Win 序，display_text 规范化。
    if (ctrl) spec.modifiers.push_back(VK_CONTROL);
    if (alt) spec.modifiers.push_back(VK_MENU);
    if (shift) spec.modifiers.push_back(VK_SHIFT);
    if (win) spec.modifiers.push_back(VK_LWIN);
    static const char* kModNames[] = {"Ctrl", "Alt", "Shift", "Win"};
    size_t name_idx = 0;
    for (UINT mod : spec.modifiers) {
        if (name_idx > 0) spec.display_text += "+";
        switch (mod) {
            case VK_CONTROL: spec.display_text += kModNames[0]; break;
            case VK_MENU: spec.display_text += kModNames[1]; break;
            case VK_SHIFT: spec.display_text += kModNames[2]; break;
            case VK_LWIN: spec.display_text += kModNames[3]; break;
            default: break;
        }
        ++name_idx;
    }
    if (!spec.display_text.empty()) spec.display_text += "+";
    spec.display_text += VkeyDisplayName(spec.vk);
    return spec;
}

} // namespace voicestick

File: desktop/macos/Sources/VoiceStickApp/KeySpec.swift
import CoreGraphics
import Foundation

/// 一次按键注入的规格（对齐 Windows key_spec.h 的 KeySpec/ParseKeySpec）：
/// 修饰键（Ctrl/Alt/Shift/Win 固定序；macOS 上映射 ⌃/⌥/⇧/⌘）+ 主键 + 规范化显示文本
/// （如 "Ctrl+Shift+V"）。供编码器旋转/单击/双击的自定义按键注入使用。
///
/// 语法：单键（up/down/left/right/enter|return/esc|escape/tab/space/backspace/delete/
/// insert/pageup/pagedown/home/end/volumeup/volumedown/volumemute/f1-f24/单字符 A-Z0-9）
/// 或修饰键组合（ctrl|control/alt/shift/win|windows|meta + 单键，"+" 分隔，
/// 大小写与前后空白不敏感）。仅修饰键、未知键名、多个主键均解析失败。
struct KeySpec: Equatable {
    /// 音量键在 macOS 上没有虚拟键码，走 NX_SYSDEFINED 媒体键事件。
    enum MediaKey: Equatable {
        case volumeUp
        case volumeDown
        case volumeMute
    }

    var control = false
    var option = false   // Alt
    var shift = false
    var command = false  // Win
    /// 普通键的 CGKeyCode；媒体键为 nil（看 mediaKey）。
    /// f21-f24 语法合法但 macOS 无对应虚拟键码：解析放行（配置跨端兼容），keyCode 为 nil，注入侧忽略。
    let keyCode: CGKeyCode?
    let mediaKey: MediaKey?
    let displayText: String

    static func parse(_ text: String) -> KeySpec? {
        var modifiers: (ctrl: Bool, alt: Bool, shift: Bool, win: Bool) = (false, false, false, false)
        var main: (keyCode: CGKeyCode?, mediaKey: MediaKey?, display: String)?

        for rawPart in text.split(separator: "+", omittingEmptySubsequences: false) {
            let part = rawPart.trimmingCharacters(in: .whitespaces).lowercased()
            if part.isEmpty { return nil }

            switch part {
            case "ctrl", "control":
                if modifiers.ctrl { return nil }
                modifiers.ctrl = true
            case "alt":
                if modifiers.alt { return nil }
                modifiers.alt = true
            case "shift":
                if modifiers.shift { return nil }
                modifiers.shift = true
            case "win", "windows", "meta":
                if modifiers.win { return nil }
                modifiers.win = true
            default:
                if main != nil { return nil }  // 多个主键
                guard let resolved = resolveMainKey(part) else { return nil }
                main = resolved
            }
        }
        guard let main else { return nil }  // 仅修饰键

        var display = ""
        if modifiers.ctrl { display += "Ctrl+" }
        if modifiers.alt { display += "Alt+" }
        if modifiers.shift { display += "Shift+" }
        if modifiers.win { display += "Win+" }
        display += main.display

        return KeySpec(
            control: modifiers.ctrl,
            option: modifiers.alt,
            shift: modifiers.shift,
            command: modifiers.win,
            keyCode: main.keyCode,
            mediaKey: main.mediaKey,
            displayText: display
        )
    }

    /// 主键名 → 键码/媒体键 + 显示名（对齐 Windows MainKeyVkey + VkeyDisplayName）。
    private static func resolveMainKey(_ lower: String) -> (CGKeyCode?, MediaKey?, String)? {
        if lower.count == 1, let char = lower.uppercased().first,
           let code = alphanumericKeyCodes[char] {
            return (code, nil, String(char))
        }
        switch lower {
        case "volumeup": return (nil, .volumeUp, "VolumeUp")
        case "volumedown": return (nil, .volumeDown, "VolumeDown")
        case "volumemute": return (nil, .volumeMute, "VolumeMute")
        default: break
        }
        if let (code, name) = namedKeys[lower] {
            return (code, nil, name)
        }
        if lower.hasPrefix("f"), lower.count >= 2 {
            let digits = lower.dropFirst()
            guard digits.allSatisfy({ $0.isNumber }), let num = Int(digits), num >= 1, num <= 24 else {
                return nil
            }
            return (functionKeyCodes[num], nil, "F\(num)")
        }
        return nil
    }

    // MARK: - 键码表（ANSI 布局）

    private static let alphanumericKeyCodes: [Character: CGKeyCode] = [
        "A": 0, "S": 1, "D": 2, "F": 3, "H": 4, "G": 5, "Z": 6, "X": 7,
        "C": 8, "V": 9, "B": 11, "Q": 12, "W": 13, "E": 14, "R": 15,
        "Y": 16, "T": 17, "1": 18, "2": 19, "3": 20, "4": 21, "6": 22,
        "5": 23, "9": 25, "7": 26, "8": 28, "0": 29, "O": 31, "U": 32,
        "I": 34, "P": 35, "L": 37, "J": 38, "K": 40, "N": 45, "M": 46
    ]

    private static let namedKeys: [String: (CGKeyCode, String)] = [
        "space": (49, "Space"),
        "enter": (36, "Enter"), "return": (36, "Enter"),
        "esc": (53, "Esc"), "escape": (53, "Esc"),
        "tab": (48, "Tab"),
        "backspace": (51, "Backspace"),
        "delete": (117, "Delete"),   // 前进删除（Windows VK_DELETE 语义）
        "insert": (114, "Insert"),   // kVK_Help，PC 键盘上即 Insert
        "up": (126, "Up"), "down": (125, "Down"),
        "left": (123, "Left"), "right": (124, "Right"),
        "pageup": (116, "PageUp"), "pagedown": (121, "PageDown"),
        "home": (115, "Home"), "end": (119, "End")
    ]

    private static let functionKeyCodes: [Int: CGKeyCode] = [
        1: 122, 2: 120, 3: 99, 4: 118, 5: 96, 6: 97, 7: 98, 8: 100,
        9: 101, 10: 109, 11: 103, 12: 111, 13: 105, 14: 107, 15: 113,
        16: 106, 17: 64, 18: 79, 19: 80, 20: 90
    ]
}
